connector/src/main/java/org/jboss/as/connector/subsystems/resourceadapters/ResourceAdapterSubsystemParser.java

```diff
@@ -194,49 +194,81 @@ private void writeRaElement(XMLExtendedStreamWriter streamWriter, ModelNode ra,
+                    streamWriter.writeStartElement(WM_SECURITY_DEFAULT_GROUPS.getXmlName());
+                    for (ModelNode group : ra.get(WM_SECURITY_DEFAULT_GROUPS.getName()).asList()) {
+                        streamWriter.writeStartElement(WM_SECURITY_DEFAULT_GROUP.getXmlName());
+                        streamWriter.writeCharacters(group.asString());
+                        streamWriter.writeEndElement();
                     }
                     streamWriter.writeEndElement();
                 }
-                if (ra.hasDefined(WM_SECURITY_MAPPING_GROUPS.getName())) {
-                    streamWriter.writeStartElement(WorkManagerSecurity.Tag.GROUPS.getLocalName());
-                    for (ModelNode node : ra.get(WM_SECURITY_MAPPING_GROUPS.getName()).asList()) {
-                        streamWriter.writeStartElement(WorkManagerSecurity.Tag.MAP.getLocalName());
-                        WM_SECURITY_MAPPING_FROM.marshallAsAttribute(node, streamWriter);
-                        WM_SECURITY_MAPPING_TO.marshallAsAttribute(node, streamWriter);
+                if (ra.hasDefined(WM_SECURITY_MAPPING_USERS.getName()) || ra.hasDefined(WM_SECURITY_MAPPING_GROUPS.getName())) {
+                    streamWriter.writeStartElement(WorkManagerSecurity.Tag.MAPPINGS.getLocalName());
+                    if (ra.hasDefined(WM_SECURITY_MAPPING_USERS.getName())) {
+                        streamWriter.writeStartElement(WorkManagerSecurity.Tag.USERS.getLocalName());
+                        for (ModelNode node : ra.get(WM_SECURITY_MAPPING_USERS.getName()).asList()) {
+                            streamWriter.writeStartElement(WorkManagerSecurity.Tag.MAP.getLocalName());
+                            WM_SECURITY_MAPPING_FROM.marshallAsAttribute(node, streamWriter);
+                            WM_SECURITY_MAPPING_TO.marshallAsAttribute(node, streamWriter);
+                            streamWriter.writeEndElement();
+                        }
+                        streamWriter.writeEndElement();
+                    }
+                    if (ra.hasDefined(WM_SECURITY_MAPPING_GROUPS.getName())) {
+                        streamWriter.writeStartElement(WorkManagerSecurity.Tag.GROUPS.getLocalName());
+                        for (ModelNode node : ra.get(WM_SECURITY_MAPPING_GROUPS.getName()).asList()) {
+                            streamWriter.writeStartElement(WorkManagerSecurity.Tag.MAP.getLocalName());
+                            WM_SECURITY_MAPPING_FROM.marshallAsAttribute(node, streamWriter);
+                            WM_SECURITY_MAPPING_TO.marshallAsAttribute(node, streamWriter);
+                            streamWriter.writeEndElement();
+                        }
                         streamWriter.writeEndElement();
                     }
                     streamWriter.writeEndElement();
                 }
                 streamWriter.writeEndElement();
+                streamWriter.writeEndElement();
```

Comment:
This conditional logic is not correct.  All defined attributes should be persisted to XML, regardless of whether or not they might be ignored at runtime.
In general, a subsystem model written to and read from XML should create operations that result in an identical model, with no loss of information.

Reply:
The intention was to add a way to persist the expression without changing the current behavior. At this moment if there is `security` element in the XML the `wm-security` is on, if there is no such element the `wm-security` is off. The `enabled` attribute is optional in case of the expression to prevent any regression. I could set the `enabled` attribute to be required instead of optional or simply use the optional attribute to hold `true` and `false` as well

In both cases, the change should be only in this class

Reply:
> The intention was to add a way to persist the expression without changing the current behavior.

A new version of an XSD is not required to retain the same syntax/semantics as an older version. The only requirement is that it must generate operations that create equivalent runtime behavior.

Since this subsystem uses a single parser class for all versions of the schema (which is a good thing!), you'll want to toggle the parsing behavior based on the version of the namespace of the schema used by the XML being parsed.

> At this moment if there is security element in the XML the wm-security is on, if there is no such element the wm-security is off.

This logic should stay in place, but only if the version of the namespace being parsed is less than the version you just created.

> The enabled attribute is optional in case of the expression to prevent any regression.

The enabled attribute is not required to be optional.  Users migrating to WF28 from WF27 will be using an older version of the schema, which will be parsed using the same logic as it was previously, and will be persisted using the new version of the schema via this subsystem's XMLElementWriter.  If the attribute is required, then the writer simply always writes its value to the XML (rather than omitting if it matches the default).

> I could set the enabled attribute to be required instead of optional or simply use the optional attribute to hold true and false as well

If this is the only attribute of the containing XML element, then it makes sense to make it required in the XSD.  However, it does seem odd that this attribute does not reside within an element that nests the configuration that it is meant to enable.  WDYT?

Reply:
I believe the problem was me trying to maintain the same behavior in the new schema and not really understanding the relation of `ResourceAdapterSubsystemParser` and `ResourceAdapterParser`. There could be a lot of simplifying if the `ResourceAdapterParser` will always persist the v7 schema and the entire conditional logic with the attribute in `ResourceAdapterSubsystemParser` can be moved into a separate method for the v7 parser. Thanks for the clarification, now I understand it better.

> However, it does seem odd that this attribute does not reside within an element that nests the configuration that it is meant to enable. WDYT?

I was thinking about making it a sub-element as well, but looking at the schema, all `enabled` properties are used as attributes, so I wanted to keep the consistency. Moreover, if the `wm-security` is set off, the entire configuration is erased. I'm afraid if there would be any kind of regression if in case of an expression there would always be one nested element. What's your view on this?

Reply:
> There could be a lot of simplifying if the ResourceAdapterParser will always persist the v7 schema

Correct.  Subsystems only register a single XML writer, which should always persist XML that adheres to the current version of the subsystem schema.

> and the entire conditional logic with the attribute in ResourceAdapterSubsystemParser can be moved into a separate method for the v7 parser.

Yes.

> I was thinking about making it a sub-element as well, but looking at the schema, all enabled properties are used as attributes, so I wanted to keep the consistency.

Actually, what you have in the XSD makes sense.
e.g.
```xml
<security enabled="...">
   <mapping-required>...</mapping-required>
   <!-- ... -->
</security>
```

What confused me is that, looking at the logic in your writer, you were ending the element after writing the WM_SECURITY attribute, but now I see that you intended only to do this when security is disabled, which we already covered on another thread.

> Moreover, if the wm-security is set off, the entire configuration is erased.

Ideally, the writer logic should be driven by what the model defines, independent of its values.  In that way, nothing should ever be "erased".  The writer should preserve all "information" in the model, such that, when re-read by the parser, the subsystem will generate an equivalent model.
Meaning, if the model for a work manager defines additional security resources/attributes, those should be persisted, even if the value of the WM_SECURITY attributes would ignore it at runtime.
This should also simply your XMLElementWriter logic immensely, I think.

> I'm afraid if there would be any kind of regression if in case of an expression there would always be one nested element.

What kind of regression do you anticipate?

Reply:
Changed the behavior as agreed, there is a new parser for the v7 schema and the `enabled` attribute is now required. 

> What kind of regression do you anticipate?

I was afraid of regression when it comes to adding another extra element, but in the end it would be solved in the parser, so bad guess on my part and it should be fine

> Ideally, the writer logic should be driven by what the model defines, independent of its values. In that way, nothing should ever be "erased". The writer should preserve all "information" in the model, such that, when re-read by the parser, the subsystem will generate an equivalent model.
Meaning, if the model for a work manager defines additional security resources/attributes, those should be persisted, even if the value of the WM_SECURITY attributes would ignore it at runtime.
This should also simply your XMLElementWriter logic immensely, I think.

I have a question regarding persisting model attributes. For example, [wm-security-mapping-required](https://docs.wildfly.org/27/wildscribe/deployment/subsystem/resource-adapters/ironjacamar/ironjacamar/resource-adapter/index.html#attr-wm-security-mapping-required) is stored in configuration and it is, as long as it's set and the `wm-security` is true (or an expression is used in this PR), but when the `wm-security` is set to false, the `wm-security-mapping-required` is set back to default value and it's no longer in the config. Is this behavior intended or should any sub-element with non-default value be persisted regardless of the `wm-security` presence in the config?

Reply:
To be precise, let's assume this situation
```
<tag>
    <wm-security enabled="true">
        <mapping-required>true</mapping-required>
    </wm-security>
</tag>
```
and I set the `wm-security` to false, should I get 
```
<tag>
    <wm-security enabled="false">
        <mapping-required>true</mapping-required>
    </wm-security>
</tag>
```
in order to persist the `mapping-required` value or can it be removed like this
```
<tag/>
```
?